src/mito2/src/schedule/scheduler.rs

```diff
@@ -0,0 +1,242 @@
+}
+
+/// Request scheduler based on local state.
+pub struct LocalScheduler {
+    /// Sends jobs to flume bounded channel
+    sender: Option<flume::Sender<Job>>,
+    /// Task handles
+    handles: Vec<Option<JoinHandle<()>>>,
+    /// Token used to halt the scheduler
+    cancel_token: CancellationToken,
+    /// State of scheduler
+    state: Arc<AtomicU8>,
+}
+
+impl LocalScheduler {
+    /// cap: flume bounded cap
+    /// num: the number of bounded receiver
+    pub fn new(cap: usize, num: usize) -> Self {
+        let (tx, rx) = flume::bounded(cap);
+        let token = CancellationToken::new();
+        let state = Arc::new(AtomicU8::new(STATE_RUNNING));
+
+        let mut handles = Vec::with_capacity(num);
+
+        for _ in 0..num {
+            let child = token.child_token().clone();
+            let receiver = rx.clone();
+            let state = Arc::clone(&state);
+            let handle = tokio::spawn(async move {
+                while state.load(Ordering::Relaxed) == STATE_RUNNING {
+                    info!("Task scheduler loop.");
+                    tokio::select! {
+                        _ = child.cancelled() => {
+                            info!("Task scheduler cancelled.");
+                            return;
+                        }
+                        req_opt = receiver.recv_async() =>{
+                            if let Ok(req) = req_opt{
+                                req.await;
+                            }
+                        }
+                    }
+                }
+                // For correctness, we need to handle all tasks
+                if state.load(Ordering::Relaxed) == STATE_AWAIT_TERMINATION {
+                    while let Ok(req) = receiver.recv() {
+                        req.await;
+                    }
+                }
+            });
+            handles.push(Some(handle));
+        }
+
+        Self {
+            sender: Some(tx),
+            cancel_token: token,
+            handles,
+            state,
+        }
+    }
+
+    #[inline]
+    fn running(&self) -> bool {
+        self.state.load(Ordering::Relaxed) == STATE_RUNNING
+    }
+}
+
+#[async_trait::async_trait]
+impl Scheduler for LocalScheduler {
+    async fn schedule(&mut self, req: Job) -> Result<()> {
+        self.sender.as_mut().unwrap().send_async(req).await.unwrap();
+        Ok(())
+    }
+
+    async fn stop(&mut self, await_termination: bool) -> Result<()> {
+        let state = if await_termination {
+            STATE_AWAIT_TERMINATION
+        } else {
+            STATE_STOP
+        };
+        self.state.store(state, Ordering::Relaxed);
+        self.cancel_token.cancel();
+        let _ = self.sender.take();
+
+        for handle in &mut self.handles {
+            if let Some(handle) = handle.take() {
+                handle.await.unwrap();
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Drop for LocalScheduler {
+    fn drop(&mut self) {
```

Comment:
Personally I don't think you should impl stop logic inside drop. It would be a bug when a scheduler is dropped without stopping it. You can check the state of scheduler and print an error log or simply panic here to warn this bug.